### ISSUE TITLE
[Roadmap] v0.2 release plan

### ISSUE THREAD

**Issue by xptree** [laekov/fastmoe#15]
# v0.2 Release

## [Feature]  Load Balance 

Implement various gating mechanisms from GShard and MoE for load balance. 
Refer to Section 2.2 of https://arxiv.org/pdf/2006.16668.pdf and https://arxiv.org/abs/1701.06538 for details. @Sengxian 

- [x] GShard's Expert capacity 
- [x] GShard's Local group dispatching
- [x] GShard's Auxiliary loss
- [x] GShard's Random routing
- [x] MoE's noisy gate

## [Enhancement] FP16 Training

- [x] FP16 backward time inspection (@laekov)

## [Feature] MoE Model Save and Load

Model save and reload for MoE with model parallel, where experts are located in different workers (@xptree)
- [x] FP32
- [x] FP16, FP16 optimizers has additional FP32 master weights to deal with.

## [Enhancement] MoELinear

- [x] Accept bias term in `MoELinear.apply` and support bias via cuBLAS's beta (@TiagoMAntunes)

## Maybe later (after v0.2)

- Support DeepSpeed, compatible to zero optimizer.

**Comment by xptree** [comment]
https://github.com/pytorch/extension-cpp/issues/5#issuecomment-385127634

**Comment by calclavia** [comment]
Any future plans to support DeepSpeed?

**Comment by gongbudaizhe** [comment]
I would recommend support ZeRO optimizer via [FairScale](https://github.com/facebookresearch/fairscale). It requires minimal code change in existing codebase to support ZeRO. DeepSpeed, on the other hand, requires a lot more code refactoring and doesn't support AMP.

**Comment by laekov** [comment]
> I would recommend support ZeRO optimizer via [FairScale](https://github.com/facebookresearch/fairscale). It requires minimal code change in existing codebase to support ZeRO. DeepSpeed, on the other hand, requires a lot more code refactoring and doesn't support AMP.

Thank you for your advise. However, v0.2.0 has already released in May, 2021. If you are looking for new features, please open a new issue.

Besides, due to a lack of full-time engineers maintaining FastMoE, we are looking forward to contributions from the community. Your contribution to support such a wonderful parallel framework will be greatly appreciated.
